Add liquidity to meta farm

kathaypacific opened the pull request:
**Don't merge yet! ** This PR updates adding and removing liquidity to use the broker bot, so that the LP's are swapped for RFP's automatically. We need to figure out what we want to do for users who already have these LPs, Charlie is going to see how many users are impacted.

This PR:
- use broker bot to add/remove liquidity
- use farmbot hook for zap and provide pages
- separate rewards from farmbot summaries, as rewards for zap and provide are calculated differently

kathaypacific committed:
not sure why the transactions are rejected

kathaypacific committed:
improve info box content

kathaypacific committed:
make meta farmbot map

kathaypacific committed:
skip fee estimate

kathaypacific committed:
Merge branch 'main' into kathy/meta-farm

kathaypacific committed:
separate rewards from farmbot so that farmbot can be reused by meta farmbot

kathaypacific committed:
use farmbot hook in liquidity card

kathaypacific committed:
add remove liquidity untested

kathaypacific committed:
update broker bot address

kathaypacific committed:
tweaks

kathaypacific committed:
update approve address to broker bot address

kathaypacific committed:
remove workarounds

jophish commented on src/pages/Compound/useFarmBotRegistry.ts:
```diff
@@ -6,39 +6,29 @@ import { AbiItem } from 'web3-utils'
-export const farmBotAddresses = ['0xCB34fbfC3b9a73bc04D2eb43B62532c7918d9E81']
-
-export const useFarmBotRegistry = () => {
+export const useFarmBotRegistry = (farmBotAddresses: string[]) => {
```
I'm getting a runtime error when running this locally on the Zap page; says that `farmBotAddresses` isn't iterable...

kathaypacific replied:
oh no :( thanks for picking this up Joe! i did miss 2 components, the old compound component (unused) and the currency list - i guess this error must have been from there. i've updated! i forgot to check this scenario before, so glad you caught it

kathaypacific replied:
i change my mind, the farmbot isn't needed on the currency list anymore (think this was used for the hackathon before we changed the UI) so i just removed the farmbot from there

jophish replied:
This is something we'll probably want again later at some point (for "zap betweens") but I think we can safely ignore that for the time being.

jophish replied:
Also, what was the rationale behind parameterizing this by a list of `farmBotAddresses`? That both `useFarmBotRegistry` and `useFarmBotRewards` now need to access a list of `farmBotAddresses`? I always kind of considered this method a sort of black box that you could call without any prior knowledge and just get all the farmbots that exist. I assume that `farmBotAddresses` will always be the same for all calls to both `useFarmBotRegistry` and `useFarmBotRewards`; could we just have both functions read from the same (shared) hardcoded list, and drop the `farmBotAddresses` parameter?

kathaypacific committed:
add farmbot address

kathaypacific committed:
remove unused

jophish commented on src/pages/Compound/useFarmBotRewards.ts:
```diff
@@ -0,0 +1,48 @@
+  tvlUSD?: string
+}
+
+export const useFarmBotRewards = (farmBotAddresses: string[]) => {
```
Why the decision to split farmbot rewards info into a new method? All farmbots have reward information associated with them. It looks like we're not getting reward info for metafarms. APY calculations for metafarms should be derived from the `FarmBotRewards` information of both the meta-RFP and the underlying RFP.

kathaypacific replied:
i decided to split the farmbot info and rewards because the meta farms need a different way to calculate the rewards to include the rewards from the underlying RFPs. i was thinking something like the meta farms rewards hook could use this hook + add some logic.

i could have gone with the approach of keeping all the logic in the hook ( e.g. `useFarmBotRegistry(farmbotType)` where `farmbotType` is an enum or something) but then the hook kind of diverges on one part of the logic and has to manage 2 sets of addresses. i felt the hook was getting a bit big and scary, so thought to keep the farmbot/meta farmbot addresses with the zap/provide functionalities but i guess this is pretty subjective. do you have a strong preference?

jophish replied:
Ahh, gotcha -- that makes sense to me! I'm fine with this approach

jophish commented on src/pages/ProvideLiquidity/LiquidityCard.tsx:
```diff
@@ -72,24 +72,24 @@ export default function LiquidityCard({
     <PoolCard
-      token0={token}
-      token1={rfpToken}
+      token0={token0}
```
Is support for displaying the APY for these metafarms going to be added in a separate PR? (I'm happy to work on that also!)

kathaypacific replied:
yes! i didn't have time to do the APY calc for this but if you'd like to take it then feel free!

jophish commented on src/pages/ProvideLiquidity/index.tsx:
```diff
@@ -77,12 +79,10 @@ export default function ProvideLiquidity() {
             <RowBetween>
               <TYPE.white fontSize={14}>
-                Add liquidity for our farmbots to support our ecosystem and earn liquidity provider fees!
+                Revo automatically stakes the liquidity that you provide, so that you can earn auto compounding rewards
```
I think having a note here about how "zap-ins to farms are supported by liquidity pools; when you provide liquidity to support zap-ins, you'll earn _additional_ auto-compounding yield farming rewards on top of the rewards from the underlying farm" or something to that effect could be helpful. i also think the docs could use some reworking to really explain what this page does/how meta-farming rewards work... happy to tackle that.

kathaypacific replied:
oh yes good idea, i will reword...

jophish reviewed (commented):
Some questions/comments/thoughts... happy to work on this PR/whatever else needs updating before tomorrow.

kathaypacific committed:
update info

cajubelt committed:
feat: remove ULP (#55)

adding a hidden /add-liquidity-legacy page that lets users remove their ULP if they provided liquidity before we switched to using the broker bot

kathaypacific committed:
adds rfp token image (#56)

Co-authored-by: Charlie Andrews-Jubelt <cajubelt@users.noreply.github.com>
